Add PR template (SQWG trial). (#394)
ros2/rcl: .github/PULL_REQUEST_TEMPLATE.md

<!--
Thanks for submitting a Pull Request!

> Please shortly explain your contribution, and if fixing an issue from the tracker, please add a link to the issue.

Reviewer: Instead of "add a link", it would be better to use the GitHub PR->issue automatic tracking.  Something like:

`if fixing an issue from the tracker, please add "Fixes #XXX", replacing XXX with the issue number`

Be sure to go over each item in the list below before submitting your pull request.
-->

### Description

Add your description here

### Checklist

> - [ ] Did you change how any of these packages work? Added new functionality? Make sure to update the documentation

Reviewer: To keep the "voice" consistent, I'd go with something more along the lines of:

```
- [ ] If you have changed how any of the APIs worked, or added new APIs, please update the API documentation.
```

> - [ ] If you added a new package: make sure to add documentation for it

Reviewer: I'd be a bit more specific on where to add documentation, something like:

Additionally, I'd add another bullet along the lines of:

```
- [ ] If you've added a new feature or new package, make sure to update the documentation at https://index.ros.org/doc/ros2/ describing the new feature.
```

> - [ ] If needed, please include (updated) tests with your PR

Reviewer: Here I would say something more along the lines of:

```
- [ ] Add or update tests corresponding to your PR.  If tests cannot be added for some reason, be sure to explain why in the description.
```

<!--
Thanks for submitting a Pull Request!

Please shortly explain your contribution, and if fixing an issue from the tracker, please add "Fixes #XXX", replacing XXX with the issue number.

Be sure that your contribution follows the [Developer Guide](https://index.ros.org/doc/ros2/Contributing/Developer-Guide/).

Be sure to go over each item in the list below before submitting your pull request.
-->

### Description

Briefly describe what your changes are doing. It is more important to mention why the changes are necessary, rather than the how.

### Checklist

- [ ] The commit messages include a [DCO](https://discourse.ros.org/t/starting-to-enforce-developer-certificate-of-origin-dco-for-some-ros-2-repos/7420).
- [ ] The commit messages follow [good practices](https://chris.beams.io/posts/git-commit/).

> - [ ] The build passes tests locally.

Reviewer: ```suggestion
- [ ] The changes passes tests locally.
```

- [ ] The PR is minimal and any unrelated changes are submitted separately.
- [ ] The API documentation has been updated based on changes in APIs or addition of new APIs, if any.
- [ ] The [documentation](https://index.ros.org/doc/ros2/) has been updated to describe the new feature or package added, if any.

> - [ ] Added or updated tests corresponding to the PR. If tests cannot be added for some reason, it has been explained in the description.

Reviewer: ```suggestion
- [ ] Tests have been added or updated to cover the changes. If tests cannot be added for some reason, it has been explained in the description.
```

> - [ ] While waiting for someone to review your PR, please consider reviewing [another open pull request](https://github.com/ros2/rcl/pulls) to support the maintainers of RCL. Refer to the [ROS code review guide](https://github.com/rosin-project/ros_code_review_guide/blob/master/README.md) for general reviewing guidelines and [Quality Guide: Ensuring code quality](https://index.ros.org/doc/ros2/Contributing/Quality-Guide/) for ROS 2 specific guidelines

Reviewer: I'd think we should probably take this out of the checklist and just make it a comment after the block.  It's not really an actionable thing on *this* PR.

Reviewer: This comment is still open.

Reviewer 2: Yes, I was just discussing this with the sqwg group once.

### Testing

> PRs must be tested on each Tier 1 platform as documented by the [REP-2000]( http://www.ros.org/reps/rep-2000.html).

Reviewer: ```suggestion
PRs must be tested on each Tier 1 platform as documented by the [REP-2000](http://www.ros.org/reps/rep-2000.html).
```

> - [ ] Linux Ubuntu x86-64

Reviewer: ```suggestion
- [ ] Linux Ubuntu amd64
```

> - [ ] Linux Ubuntu arch64

Reviewer: ```suggestion
- [ ] Linux Ubuntu arm64
```

> - [ ] OS X Testing

Reviewer: ```suggestion
- [ ] MacOS Sierra
```

> - [ ] Windows Testing

Reviewer: ```suggestion
- [ ] Windows 10 (VS2019)
```

### Release Engineering

> - [ ] This PR shoub be backported to supported ROS 2 releases. I understand changes that break API or ABI are not generally eligible to be backported.

Reviewer: ```suggestion
- [ ] This PR should be backported to supported ROS 2 releases. I understand changes that break API or ABI are not generally eligible to be backported.
```

<!--
Thanks for submitting a Pull Request!

Please shortly explain your contribution, and if fixing an issue from the tracker, please add "Fixes #XXX", replacing XXX with the issue number.

> Be ensure that your contribution follows the [Developer Guide](https://index.ros.org/doc/ros2/Contributing/Developer-Guide/).

Reviewer: ```suggestion
Be sure that your contribution follows the [Developer Guide](https://index.ros.org/doc/ros2/Contributing/Developer-Guide/).
```

Be sure to go over each item in the list below before submitting your pull request.
-->

### Description

> Add your description here

Reviewer: ```suggestion
Add your description here
```

Briefly describe what your changes are doing.  It is more important to mention why the changes are necessary, rather than the how.

### Checklist

> - [ ] Make sure that your commit messages include a [DCO]( https://discourse.ros.org/t/starting-to-enforce-developer-certificate-of-origin-dco-for-some-ros-2-repos/7420).

Reviewer: ```suggestion
- [ ] Make sure that your commit messages include a [DCO](https://discourse.ros.org/t/starting-to-enforce-developer-certificate-of-origin-dco-for-some-ros-2-repos/7420).
```

> - [ ] Make sure your commit messages follow [good practices]( https://chris.beams.io/posts/git-commit/).

Reviewer: ```suggestion
- [ ] Make sure your commit messages follow [good practices](https://chris.beams.io/posts/git-commit/).
```

- [ ] Make sure your PR is minimal. Unrelated changes should be submitted separately.
- [ ] If you have changed how any of the APIs worked, or added new APIs, please update the API documentation.

> - [ ] If you've added a new feature or new package, make sure to update the documentation at https://index.ros.org/doc/ros2/ describing the new feature.

Reviewer: ```suggestion
- [ ] If you've added a new feature or new package, make sure to update the [documentation](https://index.ros.org/doc/ros2/),  describing the new feature.
```

- [ ] Add or update tests corresponding to your PR.  If tests cannot be added for some reason, be sure to explain why in the description.
- [ ] While waiting for someone to review your PR, please consider reviewing [another open pull request](https://github.com/ros2/rcl/pulls) to support the maintainers of RCL. Refer to the [ROS code review guide](https://github.com/rosin-project/ros_code_review_guide/blob/master/README.md) for general reviewing guidelines and [Quality Guide: Ensuring code quality](https://index.ros.org/doc/ros2/Contributing/Quality-Guide/) for ROS 2 specific guidelines

> 

Reviewer: As @dirk-thomas pointed out earlier, we should also have a bullet (early in the list) about passing tests locally.

### Testing

PRs must be tested on each Tier 1 platform as documented by the [REP-2000]( http://www.ros.org/reps/rep-2000.html).

Tier 1:

- [ ] Linux Ubuntu x86-64
- [ ] Linux Ubuntu arch64
- [ ] OS X Testing
- [ ] Windows Testing

> Other tested Tier 2 and Tier 3 platforms:

Reviewer: Just remove this, since it is an empty section.

### Release Engineering

> - Should this PR cherry-picked to the next patch release? Yes/No (choose No if unsure, only possible for PRs preserving ABI compatibility)

Reviewer: ```suggestion
- [ ] Should this PR be backported to supported ROS 2 releases?  Changes that break API or ABI are not generally eligible to be backported.
```

Be sure that your contribution follows the [Developer Guide](https://index.ros.org/doc/ros2/Contributing/Developer-Guide/).

Be sure to go over each item in the list below before submitting your pull request.
-->

### Description

Briefly describe what your changes are doing. It is more important to mention why the changes are necessary, rather than the how.

### Checklist

- [ ] The commit messages include a [DCO](https://discourse.ros.org/t/starting-to-enforce-developer-certificate-of-origin-dco-for-some-ros-2-repos/7420).
- [ ] The commit messages follow [good practices](https://chris.beams.io/posts/git-commit/).
- [ ] Tests have been added or updated to cover the changes. If tests cannot be added for some reason, it has been explained in the description.
- [ ] The changes passes tests locally (see [here](https://index.ros.org/doc/ros2/Tutorials/Colcon-Tutorial) for how to run tests.
- [ ] The PR is minimal and any unrelated changes are submitted separately.
- [ ] The API documentation has been updated based on changes in APIs or addition of new APIs, if any.

> - [ ] The [documentation](https://index.ros.org/doc/ros2/) has been updated to describe the new feature or package added, if any.

Reviewer 3: For many kind of changes https://index.ros.org/doc/ros2/ isn't relevant. So this might confuse a contributor.

Reviewer 2: Should we eliminate this or provide something else along with it?

- Reviewing pull requests takes time and effort. While we are reviewing your pull request, please consider supporting the community by reviewing another open pull request. Refer to the [ROS code review guide](https://github.com/rosin-project/ros_code_review_guide/blob/master/README.md) for general reviewing guidelines and [Quality Guide: Ensuring code quality](https://index.ros.org/doc/ros2/Contributing/Quality-Guide/) for ROS 2 specific guidelines.
  - [ ] I will review (or have reviewed) another person’s PR. PR tracker No.: ___________________
  - [ ] I’m willing to review another person’s PR and would like a suggestion.

>   - [ ] I do not want to review another person’s PR.

Reviewer 4: Is this something that could be put in a comment rather than in the PR body?  I appreciate the idea of getting other people to contribute to reviewing PRs, but I would be concerned that this would be a lot of noise when going through many PRs.

Reviewer 5: > Is this something that could be put in a comment rather than in the PR body? I appreciate the idea of getting other people to contribute to reviewing PRs, but I would be concerned that this would be a lot of noise when going through many PRs.

We discussed this in the ROS Quality Working group and we thought that static comments tend to be ignored by developers. Hence, we suggested something actionable to incite the contributor to participate in reviewing other PRs.

Reviewer 6: It should be mentioned that a similar pattern is already used by other repositories, most notably [ros-planning/moveit](https://github.com/ros-planning/moveit/blob/master/.github/PULL_REQUEST_TEMPLATE.md).

### Testing

PRs must be tested on each Tier 1 platform as documented by the [REP-2000](http://www.ros.org/reps/rep-2000.html).

- [ ] Linux Ubuntu amd64
- [ ] Linux Ubuntu arm64
- [ ] MacOS Sierra

> - [ ] Windows 10 (VS2019)

Reviewer 3: Since a contributor creating a pull request doesn't have the ability to influence what platforms the CI systems checks I am not sure what the purpose of this list would be?

Reviewer 2: @thomas-moulard Could you please comment on this?

Reviewer 7: Maybe that should be changed. I would add a comment to say something that "Only contributions passing CI on all Tier 1 platform will be merged".
Some people may not realize that ROS 2 needs to compiled on Windows and OS X too, setting this expectation upfront could be helpful?

Reviewer 2: @thomas-moulard @dirk-thomas Please check if the new wordings seem alright?
